samples/net/sockets/txtime/src/main.c

```diff
@@ -0,0 +1,536 @@
+static bool connected;
+
+extern int init_vlan(void);
+
+static void quit(void)
+{
+	k_sem_give(&quit_lock);
+}
+
+static void event_handler(struct net_mgmt_event_callback *cb,
+			  uint32_t mgmt_event, struct net_if *iface)
+{
+	static bool dhcpv4_done;
+
+	if (want_to_quit) {
+		k_sem_give(&run_app);
+		want_to_quit = false;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_DHCPV4)) {
+		if (mgmt_event == NET_EVENT_IPV4_DHCP_BOUND) {
+			LOG_INF("DHCPv4 bound");
+			dhcpv4_done = true;
+
+			if (connected) {
+				k_sem_give(&run_app);
+			}
+
+			return;
+		}
+
+		if (mgmt_event == NET_EVENT_IPV4_DHCP_STOP) {
+			dhcpv4_done = false;
+			return;
+		}
+	}
+
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
+		if (!connected) {
+			LOG_INF("Network connected");
+		}
+
+		connected = true;
+
+		/* Go to connected state only after DHCPv4 is done */
+		if (!IS_ENABLED(CONFIG_NET_DHCPV4) || dhcpv4_done) {
+			k_sem_give(&run_app);
+		}
+
+		return;
+	}
+
+	if (mgmt_event == NET_EVENT_L4_DISCONNECTED) {
+		if (connected == false) {
+			LOG_INF("Waiting network to be connected");
+		} else {
+			LOG_INF("Network disconnected");
+			connected = false;
+		}
+
+		k_sem_reset(&run_app);
+
+		return;
+	}
+}
+
+static void rx(struct app_data *data)
+{
+	static uint8_t recv_buf[sizeof(txtime_str)];
+	struct sockaddr src;
+	socklen_t addr_len;
+	ssize_t len = 0;
+
+	while (true) {
+		len += recvfrom(data->sock, recv_buf, sizeof(recv_buf), 0,
+				&src, &addr_len);
+		if (!(len % (100 * 1024))) {
+			LOG_DBG("Received %zd kb data", len / 1024);
+		}
+	}
+}
+
+static void tx(struct app_data *data)
+{
+	struct net_ptp_time time;
+	struct msghdr msg;
+	struct cmsghdr *cmsg;
+	struct iovec io_vector[1];
+	union {
+		struct cmsghdr hdr;
+		unsigned char  buf[CMSG_SPACE(sizeof(uint64_t))];
+	} cmsgbuf;
+	uint64_t txtime, delay, interval;
+	int ret;
+	int print_offset;
+
+	print_offset = IS_ENABLED(CONFIG_NET_SAMPLE_PACKET_SOCKET) ?
+		sizeof(struct net_eth_hdr) : 0;
+
+	interval = CONFIG_NET_SAMPLE_PACKET_INTERVAL * NSEC_PER_USEC *
+							USEC_PER_MSEC;
+	delay = CONFIG_NET_SAMPLE_PACKET_TXTIME * NSEC_PER_USEC;
+
+	io_vector[0].iov_base = (void *)txtime_str;
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_control = &cmsgbuf.buf;
+	msg.msg_controllen = sizeof(cmsgbuf.buf);
+	msg.msg_iov = io_vector;
+	msg.msg_iovlen = 1;
+	msg.msg_name = &data->peer;
+	msg.msg_namelen = data->peer_addr_len;
+
+	cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_len = CMSG_LEN(sizeof(txtime));
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = SCM_TXTIME;
+
+	LOG_DBG("Sending network packets with SO_TXTIME");
+
+	ptp_clock_get(data->clk, &time);
+	txtime = (time.second * NSEC_PER_SEC) + time.nanosecond;
+
+	snprintk(txtime_str + print_offset,
+		 sizeof(txtime_str) - print_offset, "%"PRIx64, txtime);
+	io_vector[0].iov_len = sizeof(txtime_str);
+
+	while (1) {
+		txtime += delay;
+		*(uint64_t *)CMSG_DATA(cmsg) = txtime;
+
+		ret = sendmsg(data->sock, &msg, 0);
+		if (ret < 0) {
+			if (errno != ENOMEM) {
```

Comment:
Why is ENOMEM expected here?

Reply:
> Why is ENOMEM expected here?

The application does not have much buffers configured so it might run out of them, if sending time is longer in the future and if the sending period is very small. So for this application, it is not a big issue if we run out of buffers, we can ignore this and retry later.